sky/adaptors/aws.py

```diff
@@ -85,45 +91,55 @@ def _assert_kwargs_builtin_type(kwargs):
         f'kwargs should not contain none built-in types: {kwargs}')
 
 
-@import_package
-# The LRU cache needs to be thread-local to avoid multiple threads sharing the
-# same session object, which is not guaranteed to be thread-safe.
-@_thread_local_lru_cache()
-def session():
-    """Create an AWS session."""
-    # Creating the session object is not thread-safe for boto3,
-    # so we add a reentrant lock to synchronize the session creation.
-    # Reference: https://github.com/boto/boto3/issues/1592
+def _create_aws_object(creation_fn: Callable[[], Any]):
+    """Create an AWS object.
 
-    # Retry 5 times by default for potential credential errors,
-    # mentioned in
-    # https://github.com/skypilot-org/skypilot/pull/1988
-    max_attempts = 5
+    Args:
+        creation_fn: The function to create the AWS object.
+
+    Returns:
+        The created AWS object.
+    """
     attempt = 0
     backoff = common_utils.Backoff()
-    err = None
-    while attempt < max_attempts:
+    while True:
         try:
+            # NOTE: we need the lock here to avoid thread-safety issues when
+            # creating the resource, because Python module is a shared object,
+            # and we are not sure if the code inside 'session()' or
+            # 'session().xx()' is thread-safe.
             with _session_creation_lock:
-                # NOTE: we need the lock here to avoid
-                # thread-safety issues when creating the session,
-                # because Python module is a shared object,
-                # and we are not sure the if code inside
-                # boto3.session.Session() is thread-safe.
-                return boto3.session.Session()
+                return creation_fn()
         except (botocore_exceptions().CredentialRetrievalError,
                 botocore_exceptions().NoCredentialsError) as e:
-            time.sleep(backoff.current_backoff())
-            logger.info(f'Retry creating AWS session due to {e}.')
-            err = e
             attempt += 1
-    raise err
+            if attempt >= _MAX_ATTEMPT_FOR_CREATION:
+                raise
+            time.sleep(backoff.current_backoff())
+            logger.info(f'Retry creating AWS {creation_fn} due to {e}.')
```

Comment:
- Would `f{creation_fn}` be rendered correctly?
- Do we want this to be info or debug?
- Should `{e}` use common_utils's format_exception?